### ISSUE TITLE
Use simde_mm_move_sd to simplify simde_mm_*_sd functions

### ISSUE THREAD

**Issue by nemequ** [nemequ/simde#78]
For SSE2 and above, there are a lot of simde_mm_*_sd functions that could easily be implemented using `simde_mm_move_sd`.  For example,

```c
SIMDE__FUNCTION_ATTRIBUTES
simde__m128d
simde_mm_sub_sd (simde__m128d a, simde__m128d b) {
  simde__m128d r;

#if defined(SIMDE_SSE2_NATIVE)
  r.n = _mm_sub_sd(a.n, b.n);
#elif defined(SIMDE_ASSUME_VECTORIZATION)
  r = simde_mm_move_sd(a, simde_mm_sub_pd(a, b));
#else
  r.f64[0] = a.f64[0] - b.f64[0];
  r.f64[1] = a.f64[1];
#endif

  return r;
}
```

This lets us take advantage of the vectorization that often already exists in the `simde_mm_*_pd` functions without constantly duplicating the logic for `simde_mm_move_sd` in each function.  It also greatly simplifies adding new implementations for other ISA extensions.

Basically, it's less code, should be just as fast and much less error-prone.

I already did this for SSE with the `_mm_move_ss` function; see d01896ebac4017893eb09997f206a7b048307679.  We just need to do the same thing for `_mm_move_sd`.

**Comment by himanshi18037** [comment]
Hey!  I am really interested in working on the project and therefore was looking at some issues, but could not understand much, Is there any documentation that I could read prior to handling issues other than the documentation given? Or is there any specific part that I should concentrate my reading on. Please guide me how to proceed further with this issue and in the project as well.

**Comment by mr-c** [comment]
Hello @himanshi18037 I've added some links to background material at https://github.com/nemequ/simde/blob/more_links/README.md 

The project also has a wiki at https://github.com/nemequ/simde/wiki

Do you have experience writing C/C++? Have you written code that uses SIMD intrinsics before? If not, I would do a web search for "simd tutorial" and try one of those.

**Comment by nemequ** [comment]
Also, please feel free to use the issue tracker to ask questions!  Just open up a new issue for each one and we'll try to answer ASAP.  As far as I'm concerned questions are really documentation bugs that need to be fixed, and asking questions tells us where those bugs are.

**Comment by himanshi18037** [comment]
Hello! Thank you for the encouragement, I happy to know that I can ask questions without any hesitation. So I was tried to run run-tests.c in my system, and after forking munit folder from your other repository, I am still facing issues such as " undefined reference to `munit_rand_memory" and many more. Please let me know if I am proceeding in the right direction for this issue, or do you expect me directly move to the code for _mn_move_sd.

**Comment by himanshi18037** [comment]
> Hello @himanshi18037 I've added some links to background material at https://github.com/nemequ/simde/blob/more_links/README.md
> 
> The project also has a wiki at https://github.com/nemequ/simde/wiki
> 
> Do you have experience writing C/C++? Have you written code that uses SIMD intrinsics before? If not, I would do a web search for "simd tutorial" and try one of those.

Yes, I have good command over C language and but only know some basics of SIMD intrinsics, so would that be a problem in pursuing with the project? .

**Comment by nemequ** [comment]
> So I was tried to run run-tests.c in my system, and after forking munit folder from your other repository, I am still facing issues such as " undefined reference to `munit_rand_memory" and many more.

While the SIMDe headers can be directly included and don't require a build system, the tests do.  In the `test/` directory there is a CMake-based system.  Basically, the process is:

```
cd path/to/simd/test
mkdir build
cd build
cmake .. -DCMAKE_{C,CXX}_FLAGS='-march=native'
make -j
```

There are a lot of other ways to do it which you might prefer depending on your OS, editor/IDE, and preferences.  https://preshing.com/20170511/how-to-build-a-cmake-based-project/ looks like it does a pretty good job of explaining everything if you want more information.

If compilation is successful there should be a `run-tests` executable in the `test/build` subdirectory.

> Please let me know if I am proceeding in the right direction for this issue, or do you expect me directly move to the code for _mn_move_sd.

You're headed in the right direction.  You'll definitely want to be able to test any changes you make.

> Yes, I have good command over C language and but only know some basics of SIMD intrinsics, so would that be a problem in pursuing with the project?

Not at all.  If you were already an expert it would certainly make things easier, but it's definitely not necessary.  As long as you understand the basic ideas you should be fine.

**Comment by masterchef2209** [comment]
If #77 is done, should i start working on this issue?

**Comment by nemequ** [comment]
I think @himanshi18037 is working on this one.  Is that right, @himanshi18037?  I'd hate to have people duplicating effort, especially when there is so much to do…

@masterchef2209, might I suggest a few functions from #9 instead?  That should help you get a better idea of what is required for the GSoC stuff since it's very similar, except a lot of the time you can use previous ISA extensions as a guide since a lot of the functions are just 256-bit versions of 128-bit functions in SSE/SSE2/SSE3/SSSE3/SSE4.1.

**Comment by himanshi18037** [comment]
Hey! I am working on it but I am having trouble building the test file. I wanted to clarify if the project can easily be set up on MAC OS because that is where I am stuck.

**Comment by nemequ** [comment]
Yes, SIMDe works on macOS.  What commands are you using, and what is the error message?

If you haven't installed CMake already, you can do so via [Homebrew](https://brew.sh/) with `brew install cmake`.

**Comment by nemequ** [comment]
@himanshi18037 if it would help, we can schedule an interactive session over IRC or something (Matrix, Signal, etc.) and I'll help you get started.  Things like this can be difficult over an issue tracker or e-mail, but much simpler over some sort of real-time chat.  If you'd like to do that feel free to e-mail me and we'll find a time that works for both of us.

**Comment by kunal12298** [comment]
Hey, Kritika Gupta this side, looking forward to contribute for this project but little confused on how to get started. Please guide me and let me know on which issue I should start working.
And one more thing, can I do this project on Windows?

**Comment by kritika12298** [comment]
Sorry, I by mistake wrote from my brother's Github handle.

**Comment by nemequ** [comment]
> Hey, Kritika Gupta this side, looking forward to contribute for this project but little confused on how to get started. Please guide me and let me know on which issue I should start working.

Sorry, I thought I replied to this days ago… turns out I just left it in an open tab without posting the comment.

Someone is already working on this issue, so probably the most gentle introduction right now would be to implement an AVX2 function or two (#9).  There is a guide in the wiki explaining [the process](https://github.com/nemequ/simde/wiki/Implementing-a-New-Function).

> And one more thing, can I do this project on Windows?

SIMDe compiles and works on Windows.  However, I won't be able to provide as much help with the initial setup.  You'll need to use CMake to build the tests, https://preshing.com/20170511/how-to-build-a-cmake-based-project/ seems like a decent introduction.

If you're using Visual Studio to compile you'll want to pass `-DCMAKE_C_FLAGS="/arch:AVX2" -DCMAKE_CXX_FLAGS="/arch:AVX2"` to cmake, but keep in mind that most of the issues will require you to use a different compiler (on Windows, that probably means clang or msys2) since Visual Studio doesn't support most of the relevant targets for the GSoC tasks.

Another option might be to use Windows Subsystem for Linux.  That should let you use most Linux tools, and I should be able to help with the bits that aren't Windows-specific.

Basically, you *can* do this on Windows, but I wouldn't recommend it unless you're already very comfortable with C/C++ development on Windows.

**Comment by kritika12298** [comment]
Thank you😊 for the reply
